File: ds practical/py/hashing/hash_table_using_linear_probing.py
SIZE = 10
table = [None] * SIZE

def hash_func(key):
    return key % SIZE

def insert_key(key):
    i = hash_func(key)
    for j in range(SIZE):
        idx = (i + j) % SIZE
        if table[idx] is None:
            table[idx] = key
            print(f"Inserted {key} at index {idx}")
            return
    print("Hash table full!")

def search_key(key):
    i = hash_func(key)
    for j in range(SIZE):
        idx = (i + j) % SIZE
        if table[idx] == key:
            print(f"Found {key} at index {idx}")
            return
    print("Key not found")

def delete_key(key):
    i = hash_func(key)
    for j in range(SIZE):
        idx = (i + j) % SIZE
        if table[idx] == key:
            table[idx] = None
            print(f"Deleted {key} from index {idx}")
            return
    print("Key not found")

def display():
    print("\n--- Hash Table ---")
    for i in range(SIZE):
        print(i, ":", table[i] if table[i] is not None else "EMPTY")

# --- Example Usage ---
insert_key(11)
insert_key(21)
insert_key(31)
insert_key(4)
display()
search_key(21)
delete_key(31)
display()

File: cpp/hashing/hash_table_using_linear_probing.cpp
#include <iostream>
using namespace std;

const int SIZE = 10;
int table[SIZE];
bool occupied[SIZE] = {false};

int hashFunc(int key) {
    return key % SIZE;
}

void insertKey(int key) {
    int i = hashFunc(key);
    for (int j = 0; j < SIZE; j++) {
        int idx = (i + j) % SIZE;
        if (!occupied[idx]) {
            table[idx] = key;
            occupied[idx] = true;
            cout << "Inserted " << key << " at index " << idx << endl;
            return;
        }
    }
    cout << "Hash table full!\n";
}

void searchKey(int key) {
    int i = hashFunc(key);
    for (int j = 0; j < SIZE; j++) {
        int idx = (i + j) % SIZE;
        if (occupied[idx] && table[idx] == key) {
            cout << "Found " << key << " at index " << idx << endl;
            return;
        }
    }
    cout << "Key not found\n";
}

void deleteKey(int key) {
    int i = hashFunc(key);
    for (int j = 0; j < SIZE; j++) {
        int idx = (i + j) % SIZE;
        if (occupied[idx] && table[idx] == key) {
            occupied[idx] = false;
            cout << "Deleted " << key << " from index " << idx << endl;
            return;
        }
    }
    cout << "Key not found\n";
}

void display() {
    cout << "\n--- Hash Table ---\n";
    for (int i = 0; i < SIZE; i++) {
        if (occupied[i]) cout << i << ": " << table[i] << endl;
        else cout << i << ": EMPTY\n";
    }
}

int main() {
    insertKey(11);
    insertKey(21);
    insertKey(31);
    insertKey(4);
    display();
    searchKey(21);
    deleteKey(31);
    display();
    return 0;
}
